### PR TITLE
Feature/display happiness fixed

### LINKED ISSUE #1990: [Suggestion] Show cause of unhappiness in the townhall
### Minecolonies version
1.12.2-0.8.6705

I can't figure out what is causing my happiness not to rise in my colony. It would be nice if I could go to the town hall to find out why my colonists aren't happy. The colonists could report every day what they are not liking. Thank you.

### PR DESCRIPTION
**Pull request by TheYoxy**
Closes #1990 
# Changes proposed in this pull request:
- Adding an happiness indicator in the towhall to display every happiness factor

### CODE REVIEW

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/2509/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java [hunk 1]
```diff
@@ -994,18 +1024,24 @@ public void onUpdate()
         super.onUpdate();
 
         final String currentPage = findPaneOfTypeByID(VIEW_PAGES, SwitchView.class).getCurrentView().getID();
-        if (currentPage.equals(PAGE_PERMISSIONS))
-        {
-            updateUsers();
-            window.findPaneOfTypeByID(LIST_USERS, ScrollingList.class).refreshElementPanes();
-        }
-        else if (currentPage.equals(PAGE_CITIZENS))
+        switch (currentPage)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:1027]: CRITICAL Add a default case to this switch. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)

File: src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java [hunk 2]
```diff
@@ -100,9 +108,9 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
 
     @Override
     public void sendPackets(
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:110]: MAJOR Refactor this method to reduce its Cognitive Complexity from 23 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 3]
```diff
@@ -984,6 +984,7 @@ public static void onWorldLoad(@NotNull final World world)
                             if (colonyData != null)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:984]: CRITICAL Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)

**Review by marchermans — APPROVED** [review]

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 65 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 60 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L49](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L49): class "WindowTownHall" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L894](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L894): Reduce this anonymous class number of lines from 41 to at most 40, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [ColonyManager.java#L62](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L62): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 38 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [ColonyManager.java#L62](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L62): class "ColonyManager" has 58 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [ColonyManager.java#L353](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L353): Provide multiple methods instead of using "canDestroy" to determine which action to take. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2301)
1. ![MAJOR][MAJOR] [ColonyManager.java#L618](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L618): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [ColonyManager.java#L956](https://github.com/Minecolonies/minecolonies/blob/23c68b7d35a99e1dc145c6de135b7a990d7c9c27/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L956): Refactor this method to reduce its Cognitive Complexity from 31 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**